website/src/documentation/io/developing-io-java.md

```diff
@@ -0,0 +1,419 @@
+and why to avoid exposing your sources.
+
+### Convenience Source and Reader base classes
+
+The Beam SDK contains some convenient abstract base classes to help you create
+`Source` and `Reader` classes that work with common data storage formats, like
+files.
+
+#### FileBasedSource
+
+If your data source uses files, you can derive your `Source` and `Reader`
+classes from the `FileBasedSource` and `FileBasedReader` abstract base classes.
+`FileBasedSource` is a bounded source subclass that implements code common to
+Beam sources that interact with files, including:
+
+  * File pattern expansion
+  * Sequential record reading
+  * Split points
+
+
+## Implementing the FileBasedSink interface
+
+If your data source uses files, you can use the `FileBasedSink` interface to
+create a file-based sink for your pipeline. If your source is not file-based,
+implement a composite `PTransform` that uses `ParDo`  as discussed in the
+[developing I/O connectors overview]({{ site.baseurl }}/documentation/io/developing-io-overview/)
+
+When using the `FileBasedSink` interface, you must provide the format-specific
+logic that tells the runner how to write bounded data from your pipeline's
+`PCollection`s to an output sink. The runner writes bundles of data in parallel
+using multiple workers.
+
+Supply the logic for your file-based sink by implementing the following classes:
+
+  * A subclass of the abstract base class `FileBasedSink`. `FileBasedSink`
+    describes a location or resource that your pipeline can write to in
+    parallel.
+
+  * A user-facing wrapper `PTransform` that, as part of the logic, calls
+    [WriteFiles](https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java)
+    and passes your `FileBasedSink` as a parameter.
+
+### Implementing the FileBasedSink subclass
+
+You must create a subclass of the abstract base class `FileBasedSink`. To avoid
+exposing your sink to end-users, your class should be protected or private.
+
+The `FileBasedSink` abstract base class implements code common to Beam sources
+that interact with files, including:
+
+  * Setting file headers and footers
+  * Sequential record writing
+  * Setting the output MIME type
+
+`FileBasedSink` describes a location or resource that your pipeline can write to
+in parallel, so your `FileBasedSink` subclass might contain fields such as the
+resource or file location.
+
+`FileBasedSink` and its subclasses support writing to both local files, and
+writing to files in Google Cloud Storage. For example implementations, see the
+Beam-provided `FileBasedSink` connectors for
+[TextSink](https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/io/TextSink.java) and
+[AvroSink](https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/io/AvroSink.java).
+
+### Implementing the wrapped WriteFiles transform
```

Comment:
I think this section should be removed since following section talks about the need to wrap sources/sinks with composite transforms. Also, we mentioned WriteFiles transform before so no need to repeat here.